File: 0169-majority-element/0169-majority-element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        // USING MOORE'S VOTING ALGORITHM
        int ele,cnt = 0;
        
        for(int i = 0; i < nums.size(); i++){
            if(cnt == 0){
                ele = nums[i];
                cnt++;
            }
            
            else if(ele == nums[i]){
                cnt++;
            }
            
            else{
                cnt--;
            }
        }
        
        return ele;
    }
};


// APPROACH 2 : EXTRA SPACE 
// class Solution {
// public:
//     int majorityElement(vector<int>& nums) {
//         unordered_map<int,int> m;
//         int n = nums.size();
//         for(auto ele:nums){
//             m[ele]++;
//         }
//         for(auto ele:m){
//             if(ele.second > n/2){
//                 return ele.first;
//             }
//         }
//         return -1;
//     }
// };

File: 0169-majority-element/0169-majority-element.java
class Solution {
    public int majorityElement(int[] nums) {
        Arrays.sort(nums);
        int n = nums.length;
        
//           for(int i = 0 ;i < n; i++){
//               System.out.print(nums[i] + " ");
//           }
        
        
        int maxi = 0;
        int num = nums[0];
        int maxe = 0;
        int maxEle = nums[0];
        
        for (int i = 0; i < n; i++){

            if(num == nums[i]){
                maxe++;
            }
           else {

                if (maxe > maxi) {
                    maxi = maxe;
                    maxEle = num;
                }

                maxe = 1; 
                num = nums[i];
            }
        }
         if(maxe > maxi){
               maxEle = num;
            }
        
        return maxEle;
    }
}